glean-core/ios/Glean/Metrics/LabeledMetric.swift

```diff
@@ -0,0 +1,122 @@
+        switch subMetric {
+        case is CounterMetricType:
+            metricTypeInstantiator = glean_new_labeled_counter_metric
+        // TODO(bug XYZ): Requires the BooleanMetricType first
+        // case is BooleanMetricType:
+        //     metricTypeInstantiator = glean_new_labeled_boolean_metric
+        case is StringMetricType:
+            metricTypeInstantiator = glean_new_labeled_string_metric
+        default:
+            throw "Can not create a labeled version of this metric type"
+        }
+
+        self.handle = withArrayOfCStrings(sendInPings) { pingArray in
+            withArrayOfCStrings(labels?.sorted()) { labels in
+                metricTypeInstantiator(
+                    category,
+                    name,
+                    pingArray,
+                    Int32(sendInPings.count),
+                    lifetime.rawValue,
+                    disabled.toByte(),
+                    labels,
+                    Int32(labels?.count ?? 0)
+                )
+            }
+        }
+    }
+
+    /// Destroy this metric.
+    deinit {
+        if self.handle != 0 {
+            switch self.subMetric {
+            case is CounterMetricType:
+                glean_destroy_labeled_counter_metric(self.handle)
+            // TODO(bug XYZ): Requires the BooleanMetricType first
+            // case is BooleanMetricType:
+            //     glean_destroy_labeled_boolean_metric(self.handle)
+            case is StringMetricType:
+                glean_destroy_labeled_string_metric(self.handle)
+            default:
+                // Unreachable. The constructor will already throw an exception on an unhandled sub-metric type
+                assertUnreachable()
+            }
+        }
+    }
+
+    /// Get the specific metric for a given label.
+    ///
+    /// If a set of acceptable labels were specified in the metrics.yaml file,
+    /// and the given label is not in the set, it will be recorded under the
+    /// special `OTHER_LABEL`.
+    ///
+    /// If a set of acceptable labels was not specified in the metrics.yaml file,
+    /// only the first 16 unique labels will be used. After that, any additional
+    /// labels will be recorded under the special `OTHER_LABEL` label.
+    ///
+    /// Labels must be snake_case and less than 30 characters. If an invalid label
+    /// is used, the metric will be recorded in the special `OTHER_LABEL` label.
+    ///
+    /// - parameters:
+    ///     * label: The label
+    /// - returns: The specific metric for that label
+    subscript(label: String) -> T {
+        // swiftlint:disable force_cast
+        // REASON: We return the same type as the `subMetric` we match against
+
+        switch self.subMetric {
+        case is CounterMetricType:
+            let handle = glean_labeled_counter_metric_get(Glean.shared.handle, self.handle, label)
+            return CounterMetricType(withHandle: handle, disabled: self.disabled, sendInPings: self.sendInPings) as! T
```

Comment:
Is there ever a case where the forced cast may fail?

Reply:
Nope.
With a simple `as` cast I get an error though:
>  'CounterMetricType' is not convertible to 'T'; did you mean to use 'as!' to force downcast?

Reply:
If it's guaranteed to safely downcast then `as!` it is! Besides, it is much easier to read than than the safe cast using `as?`.